Feature/booklends

immortaldogg opened the pull request.

immortaldogg committed:
add category crud operation

- Add domain model "category"
- Add endpoints, decode, service for "category"
- Add middleware validation + error codes for "category"
- Add route for "category"
- Add sql command for creating "categories" table

immortaldogg committed:
Add CRUD for books + association between books and categories

- Can create/get/update/delete books
- Validate name/description length and category_id existence before creating a book
- Delete a category will lead to deleting all books associated with that category

immortaldogg committed:
Add booklend feature

- Add POST method to create a booklend or lend a book

fuatto commented on service/booklend/error.go:
```diff
@@ -0,0 +1,67 @@
+
+func (errInvalidTimeSpan) Error() string {
+	return `'from' must be sooner than 'to'`
+}
```
StatusCode() is missing

fuatto commented on service/booklend/pg.go:
```diff
@@ -0,0 +1,65 @@
+// IsBookAvailable implement IsBookAvailable for Booklend service
+func (s *pgService) IsBooklendable(ctx context.Context, p *domain.Booklend) (bool, error) {
+	var res = domain.Booklend{}
+	temp := s.db.Where("(booklends.from <= ? AND ? <= booklends.to) OR (booklends.from <= ? AND ? <= booklends.to)", p.From, p.From, p.To, p.To).Find(&res)
```
Check the logic

fuatto replied:
Add the book_id in query

immortaldogg committed:
Fix wrong validation on creating booklends and add error statuscode
- Add missing book_id condition in create booklend API
- Add status code to booklends error
